File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<int> stk;
        int n=heights.size();
        int res=0;

        for(int i=0;i<=n;i++){
            int curr=(i==n)?0:heights[i];

            while(!stk.empty() && heights[stk.top()]>curr){
                int h=heights[stk.top()];
                stk.pop();

                int left=stk.empty()?-1:stk.top();           //Ternary Operator
                int width=i-left-1;

                res=max(res,h*width);
            }

            stk.push(i);
        }

        return res;
    }
};

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stk=[]
        n=len(heights)
        res=0

        for i in range(n+1):
            curr=0 if i==n else heights[i]

            while(len(stk)!=0 and heights[stk[-1]]>curr):
                h=heights[stk[-1]]
                stk.pop()

                left=-1 if len(stk)==0 else stk[-1]
                width=i-left-1

                res=max(res,h*width)

            stk.append(i)

        return res